nixos/modules/system/boot/luksroot.nix

```diff
@@ -6,29 +6,38 @@ let
   luks = config.boot.initrd.luks;
 
   openCommand = name': { name, device, header, keyFile, keyFileSize, allowDiscards, yubikey, ... }: assert name' == name; ''
-    # Wait for luksRoot to appear, e.g. if on a usb drive.
-    # XXX: copied and adapted from stage-1-init.sh - should be
-    # available as a function.
-    if ! test -e ${device}; then
-        echo -n "waiting 10 seconds for device ${device} to appear..."
-        for try in $(seq 10); do
-            sleep 1
-            if test -e ${device}; then break; fi
-            echo -n .
-        done
-        echo "ok"
-    fi
+
+    # Wait for a target (e.g. device, keyFile, header, ...) to appear.
+    wait_target() {
+        local name="$1"
+        local target="$2"
+
+        if [ ! -e $target ]; then
+            echo -n "Waiting 10 seconds for $name $target to appear"
+            local success=false;
+            for try in $(seq 10); do
+                echo -n "."
+                sleep 1
+                if [ -e $target ]; then success=true break; fi
+            done
+            if [ $success = true ]; then
+                echo " - success";
+            else
+                echo " - failure";
+            fi
+        fi
+    }
+
+    # Wait for luksRoot (and optionally keyFile and/or header) to appear, e.g.
+    # if on a USB drive.
+    wait_target "device" ${device}
```

Comment:
Short answer: Yes (if we modify some stuff in `stage-1-init.sh`)

Long answer: It's a bit more complicated... Currently the commands from `luksroot.nix` are executed before `waitDevice()` is declared, i.e. we should move the code to the top (which would probably make sense for all/most functions in there). Then I would suggest that we extend the function so that we can keep the current output and the amount of seconds to wait should probably be a parameter as well (better an optional parameter, if possible). And then we would need to think about the LVM stuff (we have `preLVMCommands` and `postDeviceCommands` (post LVM) - running `lvm vgchange -ay` as pre LVM command would destroy that separation).

I thought about using `waitDevice()` but decided against it for the above reasons. Imho this commit/PR already does what it's supposed to do (imho) and improves the current situation - I hope you don't mind if I merge it :smile:.

I agree that there are a lot of possible improvements for `stage-1-init.sh` but imho that should go into a separate commit/PR - actually I already thought about some (including using `waitDevice()`) but unfortunately that might have to wait for a while since I don't have enough time for that atm.